File: day11_20/day17/lc2139_linh.py
class Solution:
    def minMoves(self, target: int, maxDoubles: int) -> int:
        ans = 0
        while target > 1:
            if maxDoubles == 0:
                ans += (target - 1)
                break
            if not target&0x1:
                target >>= 1
                maxDoubles -= 1
            else:
                target -= 1
            ans += 1
        return ans

File: docs/code/2024-11/08/lc2139_linh.ex
defmodule Solution do
  @spec min_moves(target :: integer, max_doubles :: integer) :: integer
  def min_moves(target, max_doubles) do
    move(target, max_doubles, 0)
  end

  defp move(target, 0, acc), do: acc + target - 1
  defp move(1, _, acc), do: acc
  defp move(target, max_doubles, acc) do
    case rem(target, 2) do
      0 -> move(div(target, 2), max_doubles - 1, acc + 1)
      1 -> move(target - 1, max_doubles, acc + 1)
    end
  end
end
